### ISSUE TITLE
The switch to Symfony Messenger is bad! #overhead

### ISSUE THREAD

**Issue by kortwotze** [Sylius/ShopApiPlugin#444]
## Exposé
The switch to symfony messenger brought a big issue to the plugin in terms of customizing behaviour and override classes.

### Example:
I want to change the Logic behind the POST shop-api/{channel}/carts Endpoint.

Goal is to to retrieve the last cart of a customer, if the request has provided a valid JWT and we can get a customer from the tokenStorage.

This is done in the `\Sylius\ShopApiPlugin\Handler\Cart\PickupCartHandler` class. Since we are changing the behavior instead of extending the behavior, we have to register a new service `our_overridden.shop_api.handler.cart.pickup_cart_handler`, handle dependency injection and tag it with 
`<tag name="messenger.message_handler" />`

Now we have a handler that can handle a message with type `Sylius\ShopApiPlugin\Command\Cart\PickupCart`.

From the `Sylius\ShopApiPlugin\Controller\Cart\PickupCartAction` we get a `\Sylius\ShopApiPlugin\Request\Cart\PickupCartRequest` which in turn generates the `\Sylius\ShopApiPlugin\Command\Cart\PickupCart` which is now handled by *two* handlers.
Yes, thats right - *two* handlers. 

1) `\Sylius\ShopApiPlugin\Handler\Cart\PickupCartHandler`
2) `\Our\OverriddenHandler\Cart\PickupCartHandler`

In this case, symfony messenger will deliver the message to both handlers and in this specific case, we will get the same cart persisted twice (both handlers ->add($cart)).

## Solution
This is the ugly part. We have to override *each* of the classes mentioned above to achieve a solution.

1) Tag `our_overridden.shop_api.handler.cart.pickup_cart_handler` with `            <tag name="messenger.message_handler" handles="Our\Overridden\Command\Cart\PickupCart"/>` so it handles *only that* command and not any other commands typed like that.
2) In order to do so, we have to override the `\Sylius\ShopApiPlugin\Request\Cart\PickupCartRequest` as well, to change the namespace of the PickupCart object spawned in there.
3) We have to make sure the `PickupCart` class used, does **not** extend the Sylius command object, otherwise the handler from Sylius will also trigger. We have to copy the whole class.
4) To make this work, we have to replace `\Sylius\ShopApiPlugin\Controller\Cart\PickupCartAction` as well, so we can use the new PickupCartRequest object.
5) To be able to use the new Action, we also have to re-define the route and change the service-id of the controller handling in that route.

Also, keep in mind - most overrides in this example are basically just to change namespaces to make this one small change in logic possible. Everything else is just copying files plain and simple.

... now imagine, you have customzied a lot in ShopApi. I cant provide a solution here - all I can ask for is a discussion on how to handle these things. Yes, I have seen #380 but I dont think its a solution for the messenger problem.

## Easier Solution :rocket:
Thanks to @lchrusciel, @mamazu and a good night worth of sleep, we found another solution. Its a lot better, but I think the fundamental issue with extensibility of this plugin still applies.
Anway. One can solve this problem by just replacing the service in the container. 
Use `sylius.shop_api_plugin.handler.pickup_cart_handler` as a service-id for `\Our\OverriddenHandler\Cart\PickupCartHandler` and you only have *one* service handling the `Sylius\ShopApiPlugin\Command\Cart\PickupCart` (which you obviously dispatch in your action, if you have overridden that).

**Comment by kortwotze** [comment]
I just noticed, the extension for the service tag is optional, since apparently symfony messenger handles it properly. This might be due to several attempts of fixing this locally.

Sorry for this. Everything else still applies.

**Comment by kortwotze** [comment]
We can solve 3) by using  the specialized tag from 1) of the solution.

**Comment by lchrusciel** [comment]
Hello Peter,

thanks for the detailed report. We have decided on this change because we want to work as close as it is possible with the default Symfony stack. Workflow component is out of reach for now for us, but Messenger is a valid alternative for Tactician. Also, [Symfony/Messenger](https://github.com/Symfony/Messenger) is much more actively maintained than [ThePHPLeague/Tactician](https://github.com/thephpleague/tactician). All in total, I'm still thinking it is a good decision, but the implementation could be improved.

I would say, that explicit declaration of `handles` on the tag, just to be clear, would be a nice addition.

Nevertheless, in HandlersLocator, we can find propagation to parent classes

https://github.com/symfony/messenger/blob/b2c8e491b7decb6e034fdc468ed491f02effc573/Handler/HandlersLocator.php#L69-L77

Maybe we should discuss this feature at Messenger repo and/or propose the possibility to skip parents. However, this feature may be useful, to reduce the amount of code to duplicate in some cases, because you will not need to override our handlers, just to do some additional logic.

But what you try to achieve should be possible right now. There is no point in keeping two declarations in the container (`sylius.shop_api_plugin.handler.add_coupon_handler` and `our_overridden.shop_api.handler.cart.pickup_cart_handler`). Correct me, if I'm wrong, but declaring `Our\OverriddenHandler\Cart\PickupCartHandler` as `sylius.shop_api_plugin.handler.add_coupon_handler` should fix your problem.

**Comment by kortwotze** [comment]
Yep - this is exactly what I thought of after a good night of sleep. Right now we are replacing your services with ours by keeping the service ID and thus only one "type" of handler in the container exactly as you said.

Thanks :rocket: 

I also agree on sticking with the symfony stack and sticking with standards as tight as possible. When I wrote this yesterday, I was still trying to grasp the scope of my discovery in our setup, which is indeed not small, but we'll manage. ;)
Depending on what one did with this plugin, one would have to copy & paste full classes anway (like controllers or request-objects) for extension and based from there I took an ugly route yesterday. :D

Anyway, I would still leave this issue open because it still shows the problem with extending parts of this plugin regardless of the messenger-component. Maybe we can create more discussion here on how to solve this. Or do we want this in one of the other PRs about command-providers?
